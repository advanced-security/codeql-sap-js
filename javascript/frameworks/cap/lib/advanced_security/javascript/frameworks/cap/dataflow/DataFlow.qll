/**
 * Security-related `DataFlow::Node`s or relations between two `DataFlow::Node`s.
 */

import javascript
import advanced_security.javascript.frameworks.cap.CDS

/**
 * Methods that parse source strings into a CQL expression.
 */
class ParseSink extends DataFlow::Node {
  ParseSink() {
    exists(CdsFacade cds |
      this = cds.getMember("parse").getMember(["expr", "ref", "xpr"]).getACall().getAnArgument()
    )
  }
}

/**
 * A communication happening between `cds.Service`s. This includes:
 * 1. Ones based on REST-style API, based on `cds.Service.send`,
 * 2. Ones based on query-style API, based on `cds.Services.run`, and
 * 3. Ones based on emitting and subscribing to asynchronous event messages.
 */
abstract class InterServiceCommunication extends MethodCallNode {
  InterServiceCommunication() { this.getReceiver() instanceof ServiceInstance }

  /* TODO: Generalize UserApplicationService to include built-in services such as log and db */
  /**
   * The service that sends the request.
   */
  UserDefinedApplicationService sender;
  /**
   * The service that receives the request and handles it.
   */
  UserDefinedApplicationService recipient;

  UserDefinedApplicationService getSender() { result = sender }

  UserDefinedApplicationService getReceipient() { result = recipient }
}

class RestStyleCommunication extends InterServiceCommunication {
  RestStyleCommunication() {
    exists(HandlerRegistration registration, SrvSend srvSend |
      sender = registration.getReceiver().(ServiceInstance).getDefinition() and
      recipient = srvSend.getReceiver().(ServiceInstance).getDefinition()
    )
  }
}

class CrudStyleCommunication extends InterServiceCommunication {
  CrudStyleCommunication() {
    exists(HandlerRegistration registration, SrvRun srvRun |
      sender = registration.getReceiver().(ServiceInstance).getDefinition() and
      recipient = srvRun.getReceiver().(ServiceInstance).getDefinition()
    )
  }
}

class AsyncStyleCommunication extends InterServiceCommunication {
  AsyncStyleCommunication() {
    exists(
      HandlerRegistration emittingRegistration, HandlerRegistration orchestratingRegistration,
      SrvEmit srvEmit, InterServiceCommunicationMethodCall methodCallOnReceiver
    |
      emittingRegistration != orchestratingRegistration and
      /* The service that emits the event and the service that registers the handler are the same; it's the sender. */
      this = orchestratingRegistration and
      sender = emittingRegistration.getReceiver().(ServiceInstance).getDefinition() and
      srvEmit.asExpr().getEnclosingFunction+() = sender.getInitFunction().asExpr() and
      /* 1. match by their event name. */
      srvEmit.getEmittedEvent() = orchestratingRegistration.getAnEventName() and
      /* 2. match by their service name in cds.connect().to(). */
      srvEmit.getEmitter().getDefinition().getManifestName() =
        orchestratingRegistration.getReceiver().(ServiceInstanceFromCdsConnectTo).getServiceName() and
      recipient = methodCallOnReceiver.getReceiver().(ServiceInstance).getDefinition() and
      methodCallOnReceiver.getEnclosingFunction() = orchestratingRegistration.getHandler().asExpr()
    )
  }
}
