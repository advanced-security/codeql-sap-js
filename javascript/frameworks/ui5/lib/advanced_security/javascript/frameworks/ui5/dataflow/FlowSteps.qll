import javascript
import advanced_security.javascript.frameworks.ui5.UI5

/**
 * Step from a part of internal model to a relevant control property.
 * e.g.
 * - There is a JSON model with content `{ x: null }`,
 * - There is a user-defined custom control `C` with property `{ y: { type: "string" } }`, and
 * - The two are associated in a control declaration in a view with a binding path `<C y={/x} />`.
 *
 * Then, there is a step from the content `{ x: null }` to `{ y: { type: "string" } }`.
 */
class InternalModelContentToCustomMetadataPropertyStep extends DataFlow::SharedFlowStep {
  override predicate step(DataFlow::Node start, DataFlow::Node end) {
    exists(UI5BindingPath bindingPath |
      bindingPath.getNode() = start and
      end =
        bindingPath
            .getControlDeclaration()
            .getDefinition()
            .getMetadata()
            .getProperty(bindingPath.getPropertyName())
    )
  }
}

/**
 * This is a step in the opposite direction of the `InternalModelContentToCustomMetadataPropertyStep` above. In order to ensure that this indeed holds, we check if the internal model is set to a two-way binding mode.
 */
class CustomMetadataPropertyStepToInternalModelContent extends DataFlow::SharedFlowStep {
  override predicate step(DataFlow::Node start, DataFlow::Node end) {
    exists(UI5BindingPath bindingPath, UI5InternalModel internalModel |
      start =
        bindingPath
            .getControlDeclaration()
            .getDefinition()
            .getMetadata()
            .getProperty(bindingPath.getPropertyName()) and
      bindingPath.getNode() = end and
      /* Get the content of the internal model and check if it's two-way bound. */
      internalModel.(JsonModel).getAProperty() = end and // TODO: Generalize to UI5InternalModel
      internalModel.(JsonModel).isTwoWayBinding() // TODO: Generalize to UI5InternalModel
    )
  }
}

/** External model to a relevant control property */
class ExternalModelToCustomMetadataPropertyStep extends DataFlow::SharedFlowStep {
  override predicate step(DataFlow::Node start, DataFlow::Node end) {
    exists(UI5BindingPath bindingPath |
      bindingPath.getModel() = start and
      end =
        bindingPath
            .getControlDeclaration()
            .getDefinition()
            .getMetadata()
            .getProperty(bindingPath.getPropertyName())
    )
  }
}

/** Control metadata property being the intermediate flow node */
class CustomMetadataPropertyReadStep extends DataFlow::SharedFlowStep {
  override predicate step(DataFlow::Node start, DataFlow::Node end) {
    exists(PropertyMetadata property |
      /* Writing site -> Control property */
      exists(MethodCallNode propertyWrite | propertyWrite = property.getAWrite() |
        propertyWrite.getNumArgument() = 1 and
        start = propertyWrite.getArgument(0)
        or
        propertyWrite.getNumArgument() = 2 and
        start = propertyWrite.getArgument(1)
      ) and
      end = property
      or
      /* Control property -> Reading site */
      start = property and
      end = property.getARead()
    )
  }
}

/**
 * Step from the second argument of `setProperty` method call on a local model or a reference to that model, that is, the receiver. If the contents of the model is statically visible, then get the relevant portion of the content instead.
 *
 * e.g. Given these methods in a same controller,
 *
 * ```javascript
 * onInit: {
 *   var oModel = new JSONModel({ x: null });  // A local model
 *   this.getView().setModel(oModel);
 * }
 *
 * someHandler: {
 *   this.getView().getModel().setProperty("/x", someValue);
 * }
 * ```
 *
 * Create an edge from `someValue` to `this.getView().getModel()`. As the content of `oModel` can be statically determined, also create an edge from `someValue` to `x: null`.
 */
class LocalModelSetPropertyStep extends DataFlow::SharedFlowStep {
  override predicate step(DataFlow::Node start, DataFlow::Node end) {
    /* 1. The receiver is a reference to the local model, jump to the relevant content */
    exists(MethodCallNode setPropertyCall, CustomController controller, ModelReference modelRef |
      start = setPropertyCall.getArgument(1) and
      setPropertyCall.getMethodName() = "setProperty" and
      setPropertyCall.getReceiver().getALocalSource() = modelRef and
      modelRef.asExpr().getEnclosingFunction+() = controller.getAHandler().getFunction() and // apply TC + since `modelRef` can be inside a callback argument
      controller.getAModelReference() = modelRef and
      modelRef.isLocalModelReference() and
      modelRef = end
    )
    or
    /* 2. The receiver is a reference to the local model, jump to the model reference (receiver) itself */
    exists(
      MethodCallNode setPropertyCall, CustomController controller, ModelReference modelRef,
      UI5BindingPath bindingPath
    |
      start = setPropertyCall.getArgument(1) and
      setPropertyCall.getMethodName() = "setProperty" and
      setPropertyCall.getReceiver().getALocalSource() = modelRef and
      modelRef.asExpr().getEnclosingFunction+() = controller.getAHandler().getFunction() and
      controller.getAModelReference() = modelRef and // apply TC + since `modelRef` can be inside a callback argument
      bindingPath.getNode() = modelRef.getResolvedModel().(JsonModel).getAProperty() and
      bindingPath.getPath() =
        setPropertyCall.getArgument(0).getALocalSource().asExpr().(StringLiteral).getValue() and
      end = bindingPath.getNode()
    )
    or
    /* 3. The receiver is the local model itself, jump to the relevant content */
    exists(
      MethodCallNode setPropertyCall, CustomController controller, UI5InternalModel internalModel
    |
      start = setPropertyCall.getArgument(1) and
      setPropertyCall.getMethodName() = "setProperty" and
      setPropertyCall.getReceiver().getALocalSource() = internalModel and
      internalModel.asExpr().getEnclosingFunction+() = controller.getAHandler().getFunction() and // apply TC + since `internalModel` can be inside a callback argument
      internalModel = end
    )
    or
    /* 4. The receiver is the local model, jump to the model reference (receiver) itself */
    exists(
      MethodCallNode setPropertyCall, CustomController controller, UI5InternalModel internalModel,
      UI5BindingPath bindingPath
    |
      start = setPropertyCall.getArgument(1) and
      setPropertyCall.getMethodName() = "setProperty" and
      setPropertyCall.getReceiver().getALocalSource() = internalModel and
      internalModel.asExpr().getEnclosingFunction+() = controller.getAHandler().getFunction() and // apply TC + since `internalModel` can be inside a callback argument
      bindingPath.getNode() = internalModel.(JsonModel).getAProperty() and
      bindingPath.getPath() =
        setPropertyCall.getArgument(0).getALocalSource().asExpr().(StringLiteral).getValue() and
      end = bindingPath.getNode()
    )
  }
}

/**
 * Step from a local model or a reference to it, to a `getProperty` method call on it.
 *
 * e.g Given these methods in a same controller,
 * ```javascript
 * onInit: {
 *   var oModel = new JSONModel({ x: null });  // A local model
 *   this.getView().setModel("someModel", oModel);
 * }
 *
 * someHandler1: {
 *   var value = this.getView().getModel("someModel").getProperty("/x");
 * }
 * ```
 *
 * Establish an edge from `this.getView().getModel("someModel")` in `someHandler2` to the entire `getProperty` call in `someHandler1`. Note that `modelRefFrom` and `modelRefTo` may refer to the same `ModelReference`.
 *
 * This is a dual of `LocalModelSetPropertyStep` above.
 */
class LocalModelGetPropertyStep extends DataFlow::SharedFlowStep {
  override predicate step(DataFlow::Node start, DataFlow::Node end) {
    /* 1. The receiver is a reference to the local model, jump from the relevant content */
    exists(MethodCallNode getPropertyCall, CustomController controller, ModelReference modelRef |
      modelRef = start and
      getPropertyCall.getMethodName() = "getProperty" and
      getPropertyCall.getReceiver().getALocalSource() = modelRef and
      modelRef.asExpr().getEnclosingFunction+() = controller.getAHandler().getFunction() and
      controller.getAModelReference() = modelRef and // apply TC + since `modelRef` can be inside a callback argument
      modelRef.isLocalModelReference() and
      end = getPropertyCall
    )
    or
    /* 2. The receiver is a reference to the local model, jump from the model reference (receiver) itself */
    exists(
      MethodCallNode getPropertyCall, CustomController controller, ModelReference modelRef,
      UI5BindingPath bindingPath
    |
      start = bindingPath.getNode() and
      getPropertyCall.getMethodName() = "getProperty" and
      getPropertyCall.getReceiver().getALocalSource() = modelRef and
      modelRef.asExpr().getEnclosingFunction+() = controller.getAHandler().getFunction() and
      controller.getAModelReference() = modelRef and // apply TC + since `modelRef` can be inside a callback argument
      bindingPath.getNode() = modelRef.getResolvedModel().(JsonModel).getAProperty() and
      bindingPath.getPath() =
        getPropertyCall.getArgument(0).getALocalSource().asExpr().(StringLiteral).getValue() and
      end = getPropertyCall
    )
    or
    /* 3. The receiver is the local model itself, jump from the relevant content */
    exists(
      MethodCallNode getPropertyCall, CustomController controller, UI5InternalModel internalModel
    |
      internalModel = start and
      getPropertyCall.getMethodName() = "getProperty" and
      getPropertyCall.getReceiver().getALocalSource() = internalModel and
      internalModel.asExpr().getEnclosingFunction+() = controller.getAHandler().getFunction() and // apply TC + since `internalModel` can be inside a callback argument
      end = getPropertyCall
    )
    or
    /* 4. The receiver is the local model, jump from the model reference (receiver) itself */
    exists(
      MethodCallNode getPropertyCall, CustomController controller, UI5InternalModel internalModel,
      UI5BindingPath bindingPath
    |
      start = bindingPath.getNode() and
      getPropertyCall.getMethodName() = "getProperty" and
      getPropertyCall.getReceiver().getALocalSource() = internalModel and
      internalModel.asExpr().getEnclosingFunction+() = controller.getAHandler().getFunction() and // apply TC + since `internalModel` can be inside a callback argument
      bindingPath.getNode() = internalModel.(JsonModel).getAProperty() and
      bindingPath.getPath() =
        getPropertyCall.getArgument(0).getALocalSource().asExpr().(StringLiteral).getValue() and
      end = getPropertyCall
    )
  }
}

/**
 * Step from a local model or a reference to it, to a `getProperty` method call on it. This assumes a corresponding `setProperty` call exists on the same model and its same property.
 *
 * e.g. Given these methods in a same controller,
 * ```javascript
 * onInit: {
 *   var oModel = new JSONModel({ x: null });  // A local model
 *   this.getView().setModel("someModel", oModel);
 * }
 *
 * someHandler1: {
 *   var value = this.getView().getModel("someModel").getProperty("/x");
 * }
 *
 * someHandler2: {
 *   this.getView().getModel("someModel").setProperty("/x", someValue);
 * }
 * ```
 *
 * Establish an edge from `this.getView().getModel("someModel")` in `someHandler2` to the entire `getProperty` call in `someHandler1`. Note that `modelRefFrom` and `modelRefTo` may refer to the same `ModelReference`.
 */
/*
 * TODO:
 * 1. Generalize `ModelReference` to `ModelReference` + `UI5InternalModel`.
 */

class LocalModelGetPropertyStepWithSetProperty extends DataFlow::SharedFlowStep {
  override predicate step(DataFlow::Node start, DataFlow::Node end) {
    exists(
      MethodCallNode setPropertyCall, ModelReference modelRefFrom, MethodCallNode getPropertyCall,
      ModelReference modelRefTo
    |
      setPropertyCall.getMethodName() = "setProperty" and
      setPropertyCall.getReceiver().getALocalSource() = modelRefFrom and
      start = modelRefFrom and
      getPropertyCall.getMethodName() = "getProperty" and
      getPropertyCall.getReceiver().getALocalSource() = modelRefTo and
      end = getPropertyCall and
      /*
       * Ensure that getPropertyCall and setPropertyCall are both reading/writing from/to
       * the (1) same property of the (2) same model.
       */

      getPropertyCall.getArgument(0).getALocalSource().asExpr().getStringValue() =
        setPropertyCall.getArgument(0).getALocalSource().asExpr().getStringValue() and
      modelRefFrom.getModelName() = modelRefTo.getModelName()
    )
  }
}

class LocalModelControlMetadataStep extends DataFlow::SharedFlowStep {
  override predicate step(DataFlow::Node start, DataFlow::Node end) {
    exists(
      ModelReference modelRef, BindingPath bindingPath, Binding binding, CustomControl control,
      MethodCallNode setPropertyCall
    |
      setPropertyCall.getMethodName() = "setProperty" and
      setPropertyCall.getReceiver().getALocalSource() = modelRef and
      bindingPath = binding.getBindingPath() and
      bindingPath.asString() =
        modelRef.getModelName() + ">" +
          setPropertyCall.getArgument(0).getALocalSource().asExpr().getStringValue() and
      start = modelRef and
      control.getMetadata().getProperty(binding.getBindingTarget().asXmlAttribute().getName()) = end
    )
  }
}

class SetModelToGetModelStep extends DataFlow::SharedFlowStep {
  override predicate step(DataFlow::Node start, DataFlow::Node end) {
    exists(UI5Model modelDefinition, ModelReference modelReference |
      modelReference.getResolvedModel() = modelDefinition and
      start = modelDefinition and
      end = modelReference
    )
  }
}

class GetModelToGetPropertyStep extends DataFlow::SharedFlowStep {
  override predicate step(DataFlow::Node start, DataFlow::Node end) {
    exists(ModelReference modelReference, MethodCallNode readingMethodCall |
      readingMethodCall = modelReference.getARead() and
      start = modelReference and
      end = readingMethodCall
    )
  }
}

/**
 * A step from any node that flows into the argument of
 * `ResourceBundle.getText` to its return value (which is
 * equivalent to the method call itself modulo the data flow.)
 */
class ResourceBundleGetTextCallArgToReturnValueStep extends DataFlow::SharedFlowStep {
  override predicate step(DataFlow::Node start, DataFlow::Node end) {
    exists(MethodCallNode getTextCall, ResourceModel resourceModel |
      getTextCall.getReceiver().getALocalSource() = resourceModel.getResourceBundle() and
      getTextCall.getMethodName() = "getText" and
      start = getTextCall.getArgument(1) and
      end = getTextCall
    )
  }
}