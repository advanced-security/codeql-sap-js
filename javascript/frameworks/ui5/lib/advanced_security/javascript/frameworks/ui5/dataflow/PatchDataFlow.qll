/**
 * This file patches an incompatibility introduced into the standard data flow library between
 * class DataFlow::Configurations and `summmaryModels` added in models-as-data files, and likely
 * introduced in this PR: https://github.com/github/codeql/pull/19445/files.
 */

import javascript
import semmle.javascript.dataflow.internal.FlowSummaryPrivate
private import semmle.javascript.frameworks.data.internal.ApiGraphModels as Shared

/**
 * Holds if `path` is an input or output spec for a summary with the given `base` node.
 */
pragma[nomagic]
private predicate relevantInputOutputPath(API::InvokeNode base, AccessPath inputOrOutput) {
  exists(string type, string input, string output, string path |
    ModelOutput::resolvedSummaryBase(type, path, base) and
    ModelOutput::relevantSummaryModel(type, path, input, output, _, _) and
    inputOrOutput = [input, output]
  )
}

/**
 * Gets the API node for the first `n` tokens of the given input/output path, evaluated relative to `baseNode`.
 */
private API::Node getNodeFromInputOutputPath(API::InvokeNode baseNode, AccessPath path, int n) {
  relevantInputOutputPath(baseNode, path) and
  (
    n = 1 and
    result = Shared::getSuccessorFromInvoke(baseNode, path.getToken(0))
    or
    result =
      Shared::getSuccessorFromNode(getNodeFromInputOutputPath(baseNode, path, n - 1),
        path.getToken(n - 1))
  )
}

/**
 * Gets the API node for the given input/output path, evaluated relative to `baseNode`.
 */
private API::Node getNodeFromInputOutputPath(API::InvokeNode baseNode, AccessPath path) {
  result = getNodeFromInputOutputPath(baseNode, path, path.getNumToken())
}

private predicate summaryStep(API::Node pred, API::Node succ, string kind) {
  exists(string type, string path, API::InvokeNode base, AccessPath input, AccessPath output |
    ModelOutput::relevantSummaryModel(type, path, input, output, kind, _) and
    ModelOutput::resolvedSummaryBase(type, path, base) and
    pred = getNodeFromInputOutputPath(base, input) and
    succ = getNodeFromInputOutputPath(base, output)
  )
}

/**
 * Like `ModelOutput::summaryStep` but with API nodes mapped to data-flow nodes.
 */
private predicate summaryStepNodes(DataFlow::Node pred, DataFlow::Node succ, string kind) {
  exists(API::Node predNode, API::Node succNode |
    summaryStep(predNode, succNode, kind) and
    pred = predNode.asSink() and
    succ = succNode.asSource()
  )
}

/** Data flow steps induced by summary models of kind `value`. */
private class DataFlowStepFromSummary extends DataFlow::SharedFlowStep {
  override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
    summaryStepNodes(pred, succ, "value")
  }
}

/** Taint steps induced by summary models of kind `taint`. */
private class TaintStepFromSummary extends TaintTracking::SharedTaintStep {
  override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
    summaryStepNodes(pred, succ, "taint")
  }
}
