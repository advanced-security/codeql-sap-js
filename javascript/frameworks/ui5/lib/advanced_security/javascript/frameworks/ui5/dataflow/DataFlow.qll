import javascript
import semmle.javascript.dataflow.DataFlow as StdLibDataFlow
import advanced_security.javascript.frameworks.ui5.UI5
import advanced_security.javascript.frameworks.ui5.UI5View
import advanced_security.javascript.frameworks.ui5.RemoteFlowSources
import advanced_security.javascript.frameworks.ui5.dataflow.FlowSteps
private import PatchDataFlow

/**
 * A statically visible part of a local model's content that has a binding path referring to it in a
 * control declaration acting as an HTML injection sink.
 *
 * e.g.1. Given a JSON model `oModel` declared in a controller handler and an HTML injection sink
 * `SomeSinkControl` as:
 * ```javascript
 * let oModel = new JSONModel({ y: null });
 * ```
 * and
 * ```xml
 * <SomeSinkControl x="{/y}"/>
 * ```
 * The content `y: null` of `oModel` is recognized as an instance of this class.
 *
 * e.g.2. Given a JSON model `oModel` which gains its content from a JSON file and an HTML injection
 * sink `SomeSinkControl` as:
 * ```javascript
 * let oModel = new JSONModel("controller/contents.json");
 * ```
 * and
 * ```xml
 * <SomeSinkControl x="{/y}"/>
 * ```
 * where `controller/contents.json` contains
 * ```json
 * { "y": null }
 * ```
 * The content `y: null` of `oModel` is recognized as an instance of this class.
 */
class LocalModelContentBoundBidirectionallyToHtmlISinkControl extends DomBasedXss::Sink {
  UI5BindingPath bindingPath;
  UI5Control controlDeclaration;

  LocalModelContentBoundBidirectionallyToHtmlISinkControl() {
    exists(UI5InternalModel internalModel |
      this = bindingPath.getNode() and
      (
        this instanceof PropWrite and
        internalModel.getArgument(0).getALocalSource().asExpr() =
          this.(PropWrite).getPropertyNameExpr().getParent+()
        or
        this.asExpr() instanceof StringLiteral and
        internalModel.asExpr() = this.asExpr().getParent()
      ) and
      any(UI5View view).getAnHtmlISink() = bindingPath and
      internalModel.(JsonModel).isTwoWayBinding() and
      controlDeclaration = bindingPath.getControlDeclaration()
    )
  }

  UI5BindingPath getBindingPath() { result = bindingPath }

  UI5Control getControlDeclaration() { result = controlDeclaration }
}

module UI5PathGraph<PathNodeSig ConfigPathNode, PathGraphSig<ConfigPathNode> ConfigPathGraph> {
  private newtype TNode =
    TUI5BindingPathNode(UI5BindingPath path) or
    TDataFlowNode(DataFlow::Node node)

  class UI5PathNode extends TNode {
    DataFlow::Node asDataFlowNode() { this = TDataFlowNode(result) }

    UI5BindingPath asUI5BindingPathNode() { this = TUI5BindingPathNode(result) }

    predicate isDataFlowNode() { exists(this.asDataFlowNode()) }

    predicate isUI5BindingPathNode() { exists(this.asUI5BindingPathNode()) }

    string toString() {
      result = this.asDataFlowNode().toString()
      or
      result = this.asUI5BindingPathNode().toString()
    }

    File getFile() {
      result = this.asDataFlowNode().getFile()
      or
      result = this.asUI5BindingPathNode().getView()
    }

    predicate hasLocationInfo(
      string filepath, int startline, int startcolumn, int endline, int endcolumn
    ) {
      this.asDataFlowNode().hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
      or
      this.asUI5BindingPathNode()
          .getLocation()
          .hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
    }

    ConfigPathNode getPathNode() { result.getNode() = this.asDataFlowNode() }

    UI5PathNode getAPrimarySource() {
      if this.asDataFlowNode() instanceof LocalModelContentBoundBidirectionallyToSourceControl
      then
        result.asUI5BindingPathNode() =
          this.asDataFlowNode()
              .(LocalModelContentBoundBidirectionallyToSourceControl)
              .getBindingPath()
      else result = this
    }

    UI5PathNode getAPrimaryHtmlISink() {
      if
        this.asDataFlowNode() instanceof LocalModelContentBoundBidirectionallyToHtmlISinkControl or
        this.asDataFlowNode() instanceof UI5ExternalModel // TODO: Narrow it down to ExternalModelBoundToHtmlISinkControl
      then
        result.asUI5BindingPathNode() =
          this.asDataFlowNode()
              .(LocalModelContentBoundBidirectionallyToHtmlISinkControl)
              .getBindingPath()
        or
        /* Look for the control with the binding path associated with the model */
        result.asUI5BindingPathNode().getModel().(UI5ExternalModel) = this.asDataFlowNode() and
        result.asUI5BindingPathNode() = any(UI5View view).getAnHtmlISink()
      else result = this
    }
  }

  query predicate nodes(UI5PathNode ui5PathNode) {
    exists(ui5PathNode.asUI5BindingPathNode())
    or
    exists(ConfigPathNode pathNode |
      pathNode.getNode() = ui5PathNode.asDataFlowNode() and
      ConfigPathGraph::nodes(pathNode, _, _)
    )
  }

  private module CustomPathGraphEdges {
    /**
     * An edge from the binding path in a view to a corresponding content of an internal model. e.g.
     * Given an XML declaration of a use of the control C, and a local model `oModel` it refers to:
     * ```xml
     * <C x="{/y}"/>
     * ```
     * and
     * ```javascript
     * let oModel = new JSONModel({ y: null });
     * ```
     *
     * establishes an edge from `x="{/y}"` to `y: null`.
     *
     * c.f. `FlowSteps::InternalModelContentToCustomMetadataPropertyStep`.
     */
    predicate bindingPathToInternalModelContent(
      UI5PathNode ui5PathNodePred, UI5PathNode ui5PathNodeSucc
    ) {
      exists(UI5BindingPath bindingPath |
        bindingPath = ui5PathNodePred.asUI5BindingPathNode() and
        ui5PathNodeSucc.asDataFlowNode() = bindingPath.getNode()
      )
    }

    /**
     * An edge from a content of an internal model to the corresponding binding path in a view,
     * which makes it an edge in the opposite direction as of `bindingPathToInternalModelContent`
     * above.
     * In order to ensure that the edge indeed holds, we check if the model's binding mode is
     * declared as two-way.
     *
     * c.f. `FlowSteps::InternalModelContentToCustomMetadataPropertyStep`.
     */
    predicate internalModelContentToBindingPath(
      UI5PathNode ui5PathNodePred, UI5PathNode ui5PathNodeSucc
    ) {
      exists(UI5BindingPath bindingPath, UI5InternalModel internalModel, Node boundNode |
        bindingPath = ui5PathNodeSucc.asUI5BindingPathNode() and
        boundNode = bindingPath.getNode() and
        (
          boundNode instanceof PropWrite and
          internalModel.(JsonModel).getAProperty() = boundNode // TODO: Generalize to UI5InternalModel
          or
          boundNode.asExpr() instanceof StringLiteral and
          ui5PathNodePred.asDataFlowNode() = boundNode
        ) and
        ui5PathNodePred.asDataFlowNode() = boundNode and
        internalModel.(JsonModel).isTwoWayBinding()
      )
    }
  }

  query predicate edges(UI5PathNode ui5PathNodePred, UI5PathNode ui5PathNodeSucc) {
    /* Include all existing dataflow edges */
    exists(ConfigPathNode pathNodeFrom, ConfigPathNode pathNodeTo |
      pathNodeFrom.getNode() = ui5PathNodePred.asDataFlowNode() and
      pathNodeTo.getNode() = ui5PathNodeSucc.asDataFlowNode() and
      ConfigPathGraph::edges(pathNodeFrom, pathNodeTo, _, _)
    ) and
    /* ========= TODO: Legacy code ========= */
    /* Exclude duplicate edge from model to handler parameter */
    not exists(UI5Handler h |
      ui5PathNodePred.asDataFlowNode() = h.getBindingPath().getNode() and
      ui5PathNodeSucc.asDataFlowNode() = h.getParameter(0)
    )
    or
    /* Flow to event handler parameter through the binding argument */
    exists(UI5Handler h |
      ui5PathNodePred.asUI5BindingPathNode() = h.getBindingPath() and
      ui5PathNodeSucc.asDataFlowNode() = h.getParameter(0)
    )
    or
    /* ===================================== */
    CustomPathGraphEdges::bindingPathToInternalModelContent(ui5PathNodePred, ui5PathNodeSucc)
    or
    CustomPathGraphEdges::internalModelContentToBindingPath(ui5PathNodePred, ui5PathNodeSucc)
    or
    /* Model with contents not statically observable */
    ui5PathNodePred.asUI5BindingPathNode().getModel() = ui5PathNodeSucc.asDataFlowNode() and
    ui5PathNodePred.asUI5BindingPathNode() = any(UI5View view).getASource()
    or
    ui5PathNodeSucc.asUI5BindingPathNode().getModel() = ui5PathNodePred.asDataFlowNode() and
    ui5PathNodeSucc.asUI5BindingPathNode() = any(UI5View view).getAnHtmlISink()
  }
}

module TrackPlaceAtCallConfig implements DataFlow::ConfigSig {
  /**
   * A child element instantiation.
   */
  predicate isSource(DataFlow::Node node) { node instanceof ElementInstantiation }

  /**
   * An "extension point" exposed from a parent element instantiation to
   * register a child to itself.
   */
  predicate isSink(DataFlow::Node node) { node instanceof ControlPlaceAtCall }

  /**
   * Step from data being written and the property that is being written to.
   * Needed to express the fact that the child control is usually written to
   * a property of a parent control to establish the child-parent relationship.
   *
   * NOTE: The "extension point" exposed by the parent control can come in the
   * form of:
   *
   * 1. `new Parent({ children: [ new Child( ... ) ] })`
   * 2. `new Parent(...).children[0] = new Child( ... )`
   * 3. `new Parent(...).addChild(new Child( ... ))`
   *
   * The first and second cases are (nested) `PropWrite`s to `getArgument(0)`
   * and are easily covered. But the third case poses a new problem of identifying
   * methods that writes to the `children` property, which CodeQL can't verify
   * because its definition is in the library.
   *
   * - [X] Mark all method calls: overgeneralizes, can infer dumb things
   *       (`sap.m.FlexBox.removeItem` comes to mind), but would still work for all
   *       true cases
   * - [ ] Heuristic by prefix: not an attractive option since heuristics can fail
   */
  predicate isAdditionalFlowStep(DataFlow::Node start, DataFlow::Node end) {
    exists(DataFlow::PropWrite propWrite |
      start = propWrite.getRhs() and
      end = propWrite.getBase()
    )
    or
    exists(DataFlow::MethodCallNode maybeAddingChildAPICall |
      start = maybeAddingChildAPICall.getAnArgument() and
      end = maybeAddingChildAPICall.getReceiver()
    )
    or
    exists(DataFlow::MethodCallNode call |
      start = call.getReceiver() and
      end = call
    )
    or
    exists(DataFlow::NewNode new |
      start = new.getAnArgument() and
      end = new
    )
  }
}
