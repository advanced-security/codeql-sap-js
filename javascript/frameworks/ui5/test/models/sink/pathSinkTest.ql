/**
 * @id path-sinks
 * @name Path injection sinks
 * @kind problem
 * @problem.severity error
 */

import javascript
import semmle.javascript.security.dataflow.TaintedPathQuery
import advanced_security.javascript.frameworks.ui5.dataflow.DataFlow as UI5DataFlow

class UI5ExtPathISink extends DataFlow::Node {
  UI5ExtPathISink() { this = ModelOutput::getASinkNode("ui5-path-injection").asSink() }
}

from UI5ExtPathISink sink
select sink, sink.toString()
