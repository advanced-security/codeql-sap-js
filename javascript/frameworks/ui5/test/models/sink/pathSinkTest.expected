| sink.js:51:54:51:58 | code1 | code1 |
| sink.js:53:56:53:60 | code1 | code1 |
| sink.js:55:36:55:40 | code0 | code0 |
| sink.js:57:42:57:46 | code0 | code0 |
| sink.js:59:46:59:50 | code0 | code0 |
| sink.js:63:39:63:43 | code0 | code0 |
| sink.js:65:38:65:42 | code0 | code0 |
| sink.js:67:33:67:37 | code0 | code0 |
| sink.js:69:37:69:41 | code0 | code0 |
| sink.js:71:36:71:40 | code0 | code0 |
| sink.js:71:43:71:47 | code1 | code1 |
| sink.js:73:37:73:41 | code0 | code0 |
| sink.js:73:44:73:48 | code1 | code1 |
| sink.js:75:40:75:44 | code0 | code0 |
| sink.js:75:47:75:51 | code1 | code1 |
| sink.js:77:40:77:44 | code0 | code0 |
| sink.js:77:47:77:51 | code1 | code1 |
| sink.js:92:28:92:32 | code0 | code0 |
| sink.js:94:49:94:53 | code1 | code1 |
| sink.js:96:28:96:32 | code0 | code0 |
| sink.js:115:42:115:46 | code0 | code0 |
| sink.js:116:46:116:50 | code0 | code0 |
| sink.js:118:29:118:33 | code0 | code0 |
| sink.js:118:36:118:40 | code1 | code1 |
| sink.js:119:41:119:45 | code0 | code0 |
| sink.js:119:48:119:52 | code1 | code1 |
| sink.js:120:40:120:44 | code0 | code0 |
| sink.js:120:47:120:51 | code1 | code1 |
| sink.js:122:27:122:31 | code0 | code0 |
| sink.js:122:34:122:38 | code1 | code1 |
| sink.js:122:41:122:45 | code2 | code2 |
| sink.js:122:48:122:52 | code3 | code3 |
| sink.js:122:55:122:59 | code4 | code4 |
| sink.js:122:62:122:66 | code5 | code5 |
| sink.js:123:44:123:48 | code0 | code0 |
| sink.js:123:51:123:55 | code1 | code1 |
| sink.js:123:58:123:62 | code2 | code2 |
| sink.js:123:65:123:69 | code3 | code3 |
| sink.js:123:72:123:76 | code4 | code4 |
| sink.js:123:79:123:83 | code5 | code5 |
